### PR TITLE
Refactored for distributionally robust optimization and scenario tree

### LINKED ISSUE #8: Nontrivial extension to DRO
It is not trivial to extend this package to the distributionally robust variant of stochastic programs (e.g., the formulation of Corollary 1 in http://www.optimization-online.org/DB_FILE/2020/04/7723.pdf). A main obstacle is the Lagrangian multipliers defined for non-coupling variables (i.e., `y` in the paper). To address that, the package may need to improve the design.

### LINKED ISSUE #9: User defined variables and constraints in the dual
We need a better design to allow user-defined variables and constraints to the Lagrangian master problem **without** modifying the code.

### LINKED ISSUE #13: Multistage stochastic optimization example
@hideakiv Can you add your multistage stochastic optimization example (not the DRO one) to this repo (in `example` directory)? You can create a PR.

### PR DESCRIPTION
**Pull request by hideakiv**
#8 Nontrivial extensions to DRO is handled by converting all LagrangeDual to AbstractLagrangeDual
#9 User-defined constraints and variables are also handled by AbstractLagrangeDual and initialization - the BundleMethod.jl needs to be updated
#13 added investment.jl

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -5,6 +5,8 @@ version = "0.1.0"
 
 [deps]
 BundleMethod = "e8a6717a-069f-43d5-9ac2-a49b1ec5be7b"
+Debugger = "31a5f54b-26ea-5ae9-a837-f05ce5417438"
```
Comment (kibaekkim) [inline Project.toml]: Please remove the dependency on packages `Debugger` and `GLPK`

File: src/LagrangeDual.jl [hunk 2]
```diff
@@ -133,7 +143,7 @@ function run!(LD::LagrangeDual, optimizer)
 
     if parallel.is_root()
         # Create bundle method instance
-        bundle = LD.bundle_method(num_all_coupling_variables, num_all_blocks, solveLagrangeDual)
+        bundle = LD.bundle_method(num_all_coupling_variables, num_all_blocks, solveLagrangeDual, bundle_init)
```
Comment (kibaekkim) [inline src/LagrangeDual.jl:143]: `BundleMethod.jl` needs to be updated before approving this PR.

File: examples/investment.jl [hunk 3]
```diff
@@ -0,0 +1,288 @@
+using JuMP, Ipopt, GLPK
+using DualDecomposition
+using Random
+
+const DD = DualDecomposition
+
+"""
+    Scenario Tree
```
Comment (kibaekkim) [inline examples/investment.jl]: The functions for scenario tree look written in general, which can be re-usable. @hideakiv Can you take this part and make it as part of this package by moving the part to `src`?

**Review by kibaekkim — CHANGES_REQUESTED** [review]

File: Project.toml (hunk 1 above)
Comment (hideakiv) [reply Project.toml]: Done!

File: examples/investment.jl (hunk 3 above)
Comment (hideakiv) [reply examples/investment.jl]: I added this to ScenarioTree.jl, and included this in DualDecomposition.jl

**Review by kibaekkim — APPROVED** [review]

**Review by kibaekkim — CHANGES_REQUESTED** [review]
We will wait to merge this until `BundleMethod.jl` is ready.

**Comment by kibaekkim** [comment]
@hideakiv please add some test script for `ScenarioTree.jl` in `./test/runtests.jl`.

**Comment by hideakiv** [comment]
Done!

**Comment by kibaekkim** [comment]
@hideakiv `BundleMethod.jl` just released v0.1.4 with your implementation of setting initial point. Now you have to update your package dependency and compatibility.

**Comment by kibaekkim** [comment]
And your push failed the test.

**Comment by hideakiv** [comment]
My test runs fine when I create a testset with only the "investment" problem, but crashes when I combine with the other tests. I discovered that the crash happens while we set_coupling_variables, inside parallel.allcollect. Do you have any idea as to why this is happening?

**Comment by hideakiv** [comment]
I am getting a message "Attempting to use an MPI routine after finalizing MPICH"

**Comment by kibaekkim** [comment]
> I am getting a message "Attempting to use an MPI routine after finalizing MPICH"

Yes, I confirmed that too. I suspect that internal initialization/finalization of MPI may not work properly. I am separating out the function calls in #15. Once it is merged, we can try this PR again.

**Review by kibaekkim — APPROVED** [review]
We are good to go.

**Pull request merged**